manticore/core/cpu/abstractcpu.py

```diff
@@ -479,89 +554,98 @@ def read_bytes(self, where, size):
 
     def decode_instruction(self, pc):
         '''
-        This will decode an instruction from memory pointed by `pc` and store
-        it in self.instruction.
+        This will decode an instruction from memory pointed by `pc`
 
         :param int pc: address of the instruction
         '''
-        # No dynamic code!!! #TODO!
-
-        if issymbolic(pc):
-            raise SymbolicPCException()
-
-        if not self.memory.access_ok(pc,'x'):
-            raise InvalidPCException(pc)
-
-        #Check if instruction was already decoded
+        #No dynamic code!!! #TODO! 
+        #Check if instruction was already decoded 
         self._instruction_cache = {}
         if pc in self._instruction_cache:
             logger.debug("Intruction cache hit at %x", pc)
             return self._instruction_cache[pc]
 
+        def _read_concrete_byte(address):
+            #This reads a byte from memory ignoring permissions
+            #and concretize it if symbolic
+            c = self.memory[address]
+            if issymbolic(c):
+                assert isinstance(c, BitVec) and  c.size == 8
+                if isinstance(c, Constant):
+                    c = chr(c.value)
+                else:
+                    logger.error('Concretize executable memory %r %r', c, text )
+                    raise ConcretizeMemory(self.memory, address = pc,
+                                            size = 8 * self.max_instr_width, 
+                                            policy = 'INSTRUCTION' )
+            assert isinstance(c, str)
+            return c
+
         text = ''
         try:
-            # check access_ok
+            # Read Instruction from memory
             for i in xrange(0, self.max_instr_width):
-                c = self.memory[pc+i]
-                if issymbolic(c):
-                    assert isinstance(c, BitVec) and  c.size == 8
-                    if isinstance(c, Constant):
-                        c = chr(c.value)
-                    else:
-                        logger.error('Concretize executable memory %r %r', c, text )
-                        break
-                assert isinstance(c, str)
-                text += c
-        except MemoryException:
+                text += _read_concrete_byte(pc+i)
+        except MemoryException as mem_except:
+            #lets not raise this exception in case the instruction
+            #was fully read befor the page
             pass
-        
+
+        #Pad potentially incomplete intruction with zeroes
         code = text.ljust(self.max_instr_width, '\x00')
-        instruction = next(self._md.disasm(code, pc))
 
-        #PC points to symbolic memory 
-        if instruction.size > len(text):
-            logger.info("Trying to execute instructions from invalid memory")
-            raise InvalidPCException(pc)
+        #decode the instructtion from code 
+        try:
+            instruction = next(self._md.disasm(code, pc))
+        except Exception as e:
```

Comment:
There is a CsError. But I couldn't make it throw it. 
I got 
StopIteration, ctypes.ArgumentError/TypeError


felipe@localhost /tmp $ python ssss.py 
Traceback (most recent call last):
  File "ssss.py", line 7, in <module>
    i = next( md.disasm(CODE, "SAS"))
  File "/home/felipe/.local/lib64/python2.7/site-packages/capstone/__init__.py", line 849, in disasm
    res = _cs.cs_disasm(self.csh, code, len(code), offset, count, ctypes.byref(all_insn))
ctypes.ArgumentError: argument 4: <type 'exceptions.TypeError'>: wrong type
felipe@localhost /tmp $ python ssss.py 
Traceback (most recent call last):
  File "ssss.py", line 7, in <module>
    i = next( md.disasm(2, 0x1000))
  File "/home/felipe/.local/lib64/python2.7/site-packages/capstone/__init__.py", line 849, in disasm
    res = _cs.cs_disasm(self.csh, code, len(code), offset, count, ctypes.byref(all_insn))
TypeError: object of type 'int' has no len()
felipe@localhost /tmp $ python ssss.py 
Traceback (most recent call last):
  File "ssss.py", line 7, in <module>
    i = next( md.disasm(CODE, 0x1000,-1))
StopIteration